sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventProcessor.java

```diff
@@ -229,14 +236,48 @@ private void receiveEvents(PartitionOwnership partitionOwnership) {
         PartitionProcessor partitionProcessor = this.partitionProcessorFactory
             .createPartitionProcessor(partitionContext, checkpointManager);
         partitionProcessor.initialize().subscribe();
-
+        final AtomicReference<Context> processSpanContext = new AtomicReference<>(Context.NONE);
         consumer.receive().subscribeOn(Schedulers.newElastic("PartitionPump"))
-            .subscribe(eventData -> partitionProcessor.processEvent(eventData).subscribe(unused -> {
-            }, partitionProcessor::processError),
+            .subscribe(eventData -> {
+                startScopedTracingSpan(eventData, processSpanContext);
+                partitionProcessor.processEvent(eventData).subscribe(unused -> {
+                }, partitionProcessor::processError);
+                endScopedTracingSpan(processSpanContext);
+            },
                 partitionProcessor::processError,
                 // Currently, there is no way to distinguish if the receiver was closed because
                 // another receiver with higher/same owner level(epoch) connected or because
                 // this event processor explicitly called close on this consumer.
                 () -> partitionProcessor.close(CloseReason.LOST_PARTITION_OWNERSHIP));
     }
+
+    /*
+     * Starts a new process tracing span and attached context the EventData object for users.
+     */
+    private void startScopedTracingSpan(EventData eventData, AtomicReference<Context> processSpanContext) {
+        Object diagnosticId = eventData.properties().get(DIAGNOSTIC_ID_KEY);
+        if (diagnosticId == null) {
+            return;
+        }
+        eventData.context(TraceUtil.extractContext(diagnosticId.toString(), Context.NONE));
+        processSpanContext.set(TraceUtil.startScopedSpan("process", eventData.context()));
+        eventData.context(processSpanContext.get());
+    }
+
+    /*
+     * Ends the tracing span and the scope of that span.
+     */
+    private void endScopedTracingSpan(AtomicReference<Context> processSpanContext) {
+        // Disposes of the scope when the trace span closes.
+        if (!processSpanContext.get().getData("scope").isPresent()) {
+            return;
+        }
+        Closeable close = (Closeable) processSpanContext.get().getData("scope").get();
+        try {
+            close.close();
+        } catch (IOException ioException) {
+            logger.error("EventProcessor.run() endTracingSpan().close() failed with an error %s", ioException);
+        }
+        TraceUtil.endTracingSpan(processSpanContext.get(), null);
```

Comment:
what if processing results with an exception on the user side, how do we set error status for processing span?

Reply:
the doOnEach will get us OnComplete and OnError so then we can just pass down the error and have the user error message in the description for Status.